dashboard/app/models/authentication_option.rb

```diff
@@ -32,8 +32,8 @@ class AuthenticationOption < ApplicationRecord
   validates :email, no_utf8mb4: true
   validates_email_format_of :email, allow_blank: true, if: :email_changed?, unless: -> {email.to_s.utf8mb4?}
 
-  validate :email_must_be_unique, :hashed_email_must_be_unique,
-    :cred_type_and_auth_id_must_be_unique
+  validate :email_must_be_unique, :hashed_email_must_be_unique, unless: -> {UNTRUSTED_EMAIL_CREDENTIAL_TYPES.include? credential_type}
```

Comment:
One thing that concerns me about this approach:  Although the untrusted credential type with a non-unique email address remains valid, it can still _invalidate_ an existing trusted credential type with that same email address.

```rb
trusted = build :authentication_option, credential_type: 'google_oauth2'
assert trusted_option.valid?

untrusted = build :authentication_option, credential_type: 'clever', email: trusted.email
assert untrusted.valid?
refute trusted.valid? # this seems bad
```

Also possible I misread this because it's 6pm on a Friday.

Reply:
Oh, good catch!

Should we maybe only care about uniqueness among trusted credential types? Something like

```ruby
  private def email_must_be_unique
    # only enforce for credential types for which we trust the email
    return unless TRUSTED_EMAIL_CREDENTIAL_TYPES.include? credential_type
    # skip the db lookup if possible
    return unless email_changed? && email.present? && errors.blank?

    other = AuthenticationOption.find_by(email: email, credential_type: TRUSTED_EMAIL_CREDENTIAL_TYPES)
    if other && other != self
      errors.add :email, I18n.t('errors.messages.taken')
    end
  end
```

Note this has the side effect of also only enforcing uniqueness among migrated users.

Reply:
Yeah, that's what I was thinking; an untrusted email should neither invalidate nor be invalidated by a trusted email.  They're exclusive groups.  But then...

> Note this has the side effect of also only enforcing uniqueness among migrated users.

I am a bit worried about the chaos this could create, and the pain of digging ourselves out of it should we ever need to upgrade (say) Clever to be a "trusted email" provider.  Even if this wasn't true, and we checked emails of unmigrated users with "trusted email" providers, it seems like we could end up with a large number of duplicate emails.  Also haven't thought through what existing and future security holes that somewhat intuitive model could cause.  Do we block password resets being sent to users that don't have a "trusted email"?

I'm starting to think we'd be better off just not recording an email address at all for an "untrusted email" provider, for students or teachers.  If the teacher needs an email address on their account, they'll have to add it as an email Authentication Option (which, coincidentally, _ought_ to be untrusted but we don't treat it as such :cry:).  The ramification for our mailing lists is not great, but insofar as I want account data to represent what we actually know about the user it seems like a better approach.

Am I throwing the baby out with the bathwater?

Reply:
I started digging into this a bit more, and I think a better approach would be something like

```diff
diff --git a/dashboard/app/models/authentication_option.rb b/dashboard/app/models/authentication_option.rb
index 818837a1be5..58eaed0dd69 100644
--- a/dashboard/app/models/authentication_option.rb
+++ b/dashboard/app/models/authentication_option.rb
@@ -32,8 +32,8 @@ class AuthenticationOption < ApplicationRecord
   validates :email, no_utf8mb4: true
   validates_email_format_of :email, allow_blank: true, if: :email_changed?, unless: -> {email.to_s.utf8mb4?}
 
-  validate :email_must_be_unique, :hashed_email_must_be_unique,
-    :cred_type_and_auth_id_must_be_unique
+  validate :email_must_be_unique, :hashed_email_must_be_unique, unless: -> {UNTRUSTED_EMAIL_CREDENTIAL_TYPES.include? credential_type}
+  validate :cred_type_and_auth_id_must_be_unique
 
   after_create :set_primary_contact_info
 
@@ -48,17 +48,21 @@ class AuthenticationOption < ApplicationRecord
     TWITTER = 'twitter',
     WINDOWS_LIVE = 'windowslive',
     MICROSOFT = 'microsoft_v2_auth',
-  ]
+  ].freeze
+
+  CREDENTIAL_TYPES = [
+    EMAIL = 'email',
+    OAUTH_CREDENTIAL_TYPES,
+  ].flatten.freeze
 
   UNTRUSTED_EMAIL_CREDENTIAL_TYPES = [
     CLEVER,
     POWERSCHOOL
   ].freeze
 
-  CREDENTIAL_TYPES = [
-    EMAIL = 'email',
-    OAUTH_CREDENTIAL_TYPES,
-  ].flatten
+  TRUSTED_EMAIL_CREDENTIAL_TYPES = (
+    CREDENTIAL_TYPES - UNTRUSTED_EMAIL_CREDENTIAL_TYPES
+  ).freeze
 
   SILENT_TAKEOVER_CREDENTIAL_TYPES = [
     FACEBOOK,
@@ -66,7 +70,9 @@ class AuthenticationOption < ApplicationRecord
     # TODO: (madelynkasula) Remove once we are sure users are no longer logging in via windowslive.
     WINDOWS_LIVE,
     MICROSOFT
-  ]
+  ].freeze
+
+  scope :trusted_email, -> {where(credential_type: TRUSTED_EMAIL_CREDENTIAL_TYPES)}
 
   def codeorg_email?
     Mail::Address.new(email).domain == 'code.org'
diff --git a/dashboard/app/models/user.rb b/dashboard/app/models/user.rb
index fd03a50aae6..179219d9639 100644
--- a/dashboard/app/models/user.rb
+++ b/dashboard/app/models/user.rb
@@ -551,7 +551,7 @@ class User < ActiveRecord::Base
   # @return [User|nil]
   def self.find_by_email(email)
     return nil if email.blank?
-    migrated_user = AuthenticationOption.find_by(email: email)&.user
+    migrated_user = AuthenticationOption.trusted_email.find_by(email: email)&.user
     migrated_user || User.find_by(email: email)
   end
 
@@ -560,7 +560,7 @@ class User < ActiveRecord::Base
   # @return [User|nil]
   def self.find_by_hashed_email(hashed_email)
     return nil if hashed_email.blank?
-    migrated_user = AuthenticationOption.find_by(hashed_email: hashed_email)&.user
+    migrated_user = AuthenticationOption.trusted_email.find_by(hashed_email: hashed_email)&.user
     migrated_user || User.find_by(hashed_email: hashed_email)
   end
```

Specifically, we also update the concept of identifying a user by email to only consider trusted emails.

> I'm starting to think we'd be better off just not recording an email address at all for an "untrusted email" provider, for students or teachers

I'm sort of headed this direction myself. The diff I have here puts us in a place halfway to that, where we are still storing the emails but are using them less. It means we still have access to more information if we want to use it for the future, but also has the potential for that information to be misinterpreted moving forward so I'm not sure where I land on which I prefer.

> (which, coincidentally, ought to be untrusted but we don't treat it as such cry)

Actually, it might be worth changing our terminology here. I think there's a difference between an email being untrusted because we can't validate that the user is actually the person who has access to the email and the email being untrusted because the user has no control over which email they give us. Most of this work here is concerned with support for the latter case.

Reply:
> I think there's a difference between an email being untrusted because we can't validate that the user is actually the person who has access to the email and the email being untrusted because the user has no control over which email they give us.

Oh, that's a useful distinction.

- With the former _unvalidated_ email, we **can** enforce uniqueness because we're in control of the "point of entry" for the email address, and we **must** enforce uniqueness because we use that email as an identifier for the user.
- With the latter _suggested_ (?) email, we **cannot** enforce uniqueness because we're not in control of the "point of entry" for the email address, and we **cannot** trust uniqueness because the provider does not enforce it, therefore we **must not** enforce uniqueness, which also implies we **must not** use that email as an identifier for the user.

Reply:
In addition to your latest commit, may I suggest adding a test like the one at the top of this thread, checking interactions between authentication options re: validity?

Reply:
Done! Also added a comment going into deeper detail about what it means for an email to be "untrusted", and added a jira item [here](https://codedotorg.atlassian.net/browse/FND-898) to track investigation into encoding this distinction into our data model.